### PR TITLE
DB-11302 Performance improvement for statement triggers on Spark

### PR DESCRIPTION
**Pull request by msirek**
Performance issues fixed:
1.  Creates a persisted rdd for reading trigger referencing rows instead of reading from the temporary conglomerate each time.
2. Avoids merge join on spark if the target of an update or delete is involved.  Frequently updated tables tend to generate uneven splits unless flushed to HDFS first.  Other joins like merge sort join or broadcast join repartition the data, and create more even partitions, making these joins perform better.  More work is needed in the area of picking split points for operations which don't repartition the data set.
3. Convert the trigger rows rdd to a native spark Dataset, and apply the Dataset limit function on it when processing WHEN clause predicates.  The native spark Dataset is reused instead of building a new splice DataSet for each reference to NEW TABLE or OLD TABLE rows.

This PR should merge after https://github.com/splicemachine/spliceengine/pull/4943 because it fixes some serialization issues needed for ITs to pass.

### CODE REVIEW

**Comment by msirek** [comment]
jenkins please test branch @cdh6.3.0

**Comment by msirek** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by msirek** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2481/

**Comment by msirek** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2522/

**Comment by msirek** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2534/

**Comment by msirek** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Review by martinrupp — CHANGES_REQUESTED** [review]
this might change serialization of GenericColumnDescriptor.
have we tested that upgrade still works with this?

**Review by martinrupp — CHANGES_REQUESTED** [review]
this PR break upgrades.
```
git checkout origin/master
./start-splice-cluster
./sqlshell.sh
```
SQL:
```
create table a ( i int);
create table b ( i int);
create trigger tg2 after insert on a  REFERENCING NEW AS NEW_ROW for each row when (NEW_ROW.i = 1) begin atomic insert into b values 1; insert into b values 2; end;
create trigger tg3 after insert on b  REFERENCING NEW AS NEW_ROW for each row when (NEW_ROW.i = 3) begin atomic insert into a values 1; insert into b values 2; end;
```
checkout this change, restart

```
./start-splice-cluster -k
git checkout origin/DB-11302
./start-splice-cluster -l
./sqlshell.sh
```
call SYSCS_INVALIDATE_STORED_STATEMENTS:
```
splice> call SYSCS_UTIL.SYSCS_INVALIDATE_STORED_STATEMENTS();
ERROR SE001: Splice Engine exception: unexpected exception
ERROR XJ001: Java exception: 'com.esotericsoftware.kryo.KryoException: Encountered unregistered class ID: 71: java.io.IOException'.
ERROR XJ001: Java exception: 'Encountered unregistered class ID: 71: com.esotericsoftware.kryo.KryoException'.
```

**Comment by msirek** [comment]
> this might change serialization of GenericColumnDescriptor.
> have we tested that upgrade still works with this?

I did a simple upgrade test, and serialized items on disk can be read.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/HashableJoinStrategy.java [hunk 1]
```diff
@@ -78,15 +78,6 @@ public boolean feasible(Optimizable innerTable,
         }
         */
 
-		/* Don't consider hash join on the target table of an update/delete.
```
[the remark this answers is not in the document]
Comment (msirek) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/HashableJoinStrategy.java:81]: It seems to be an old limitation.  I tried some update tests and did not observe the issue mentioned.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/execute/TriggerExecutionContext.java [hunk 2]
```diff
@@ -854,4 +858,22 @@ public String getTriggerName() {
...
+
+    public void setHasGeneratedColumn() {
+        hasGeneratedColumn = true;
+    }
+
+    public boolean hasGeneratedColumn() {
+        return hasGeneratedColumn;
+    }
+
+    public boolean needsTemporaryConglomerate() {
```
[the remark this answers is not in the document]
Comment (msirek) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/execute/TriggerExecutionContext.java:870]: Tables with generated columns still need to read trigger rows from the temporary conglomerate, so they will not see a performance boost.

File: splice_machine/src/main/java/com/splicemachine/derby/impl/sql/compile/MergeJoinStrategy.java [hunk 3]
```diff
@@ -83,6 +94,15 @@ public boolean feasible(Optimizable innerTable,
...
         }
+
+        // Merge joins involving the target of an UPDATE or DELETE can be slow on Spark
+        // because the input splits for a table which is getting frequently updated
+        // are currently not calculated accurately, leading to uneven splits.
+        // It is safer to pick a join which will repartition the tables into multiple even
+        // partitions such as MergeSortJoin.
+        if (optimizer.isForSpark() && !wasHinted && joinHasTargetTable(innerTable, optimizer))
+            return false;
+
```
Comment (dgomezferro) [inline splice_machine/src/main/java/com/splicemachine/derby/impl/sql/compile/MergeJoinStrategy.java:105]: I think the bug here is that we are not computing good splits. Can you reference https://splicemachine.atlassian.net/browse/DB-7642 here ? Once that's fixed we should remove this restriction.

File: hbase_sql/src/main/java/com/splicemachine/stream/StreamListener.java [hunk 4]
```diff
@@ -127,7 +128,19 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
...
+            // Give the SparkException some time to get filled in to
+            // the failure field.
+            if (failure instanceof IOException) {
+                if ("Connection reset by peer".equals(failure.getMessage())) {
+                    try {
+                        Thread.sleep(100);
+                    } catch (InterruptedException e) {
+                    }
+                }
+            }
```
[the remark this answers is not in the document]
Comment (msirek) [reply hbase_sql/src/main/java/com/splicemachine/stream/StreamListener.java:143]: @dgomezferro There seems to be a race condition, as a byproduct of shutting down the operation one of the tasks gets an IOException.  Eventually the original cause gets filled in.  I didn't really have time to trace out the exact root cause.  If you wish I could remove this code if you'd like to investigate it further.   This is an IT failure currently happening on master as well that @hatyo encountered.

File: splice_machine/src/main/java/com/splicemachine/derby/impl/sql/compile/MergeJoinStrategy.java (hunk 3 above)
Comment (msirek) [reply splice_machine/src/main/java/com/splicemachine/derby/impl/sql/compile/MergeJoinStrategy.java]: Added a TODO comment referencing DB-7642

**Comment by msirek** [comment]
@martinrupp  After https://github.com/splicemachine/spliceengine/pull/4943 merges, please let me know, and I will update this PR to remove the new SerDe code it added.  Only after that would it be ready to merge.

**Comment by msirek** [comment]
jenkins please test branch @cdh6.3.0

**Comment by msirek** [comment]
@martinrupp @dgomezferro @arnaud-splice I've resolved the serialization issue after merging the code with DB-10410.  Please take a look if you get the chance.

File: hbase_sql/src/main/java/com/splicemachine/stream/StreamListener.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (msirek) [reply hbase_sql/src/main/java/com/splicemachine/stream/StreamListener.java]: I have removed this code.

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Review by martinrupp — APPROVED** [review]

**Review by dgomezferro — APPROVED** [review]

**Review by arnaud-splice — APPROVED** [review]

**Review by carolp-503 — APPROVED** [review]
+1

**Pull request merged**